vm/systemSmartContracts/validator.go

```diff
@@ -1069,46 +1075,73 @@ func (v *validatorSC) stake(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
 		}
 	}
 
-	if !v.isNumberOfNodesTooHigh(registrationData) {
-		v.activateStakingFor(
-			blsKeys,
-			registrationData,
-			validatorConfig.NodePrice,
-			registrationData.RewardAddress,
-			args.CallerAddr,
-		)
-	} else {
-		numRegisteredBlsKeys := int64(len(registrationData.BlsPubKeys))
+	v.activateNewBLSKeys(registrationData, blsKeys, newKeys, &validatorConfig, args)
+
+	err = v.saveRegistrationData(args.CallerAddr, registrationData)
+	if err != nil {
+		v.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (v *validatorSC) activateNewBLSKeys(
+	registrationData *ValidatorDataV2,
+	blsKeys [][]byte,
+	newKeys [][]byte,
+	validatorConfig *ValidatorConfig,
+	args *vmcommon.ContractCallInput,
+) {
+	numRegisteredBlsKeys := len(registrationData.BlsPubKeys)
+	numNodesTooHigh := v.activateStakingFor(
+		blsKeys,
+		newKeys,
+		registrationData,
+		validatorConfig.NodePrice,
+		registrationData.RewardAddress,
+		args.CallerAddr,
+	)
+
+	if numNodesTooHigh && len(blsKeys) > 0 {
 		nodeLimit := int64(v.computeNodeLimit())
 		entry := &vmcommon.LogEntry{
 			Identifier: []byte(args.Function),
 			Address:    args.RecipientAddr,
 			Topics: [][]byte{
 				[]byte(numberOfNodesTooHigh),
-				big.NewInt(numRegisteredBlsKeys).Bytes(),
+				big.NewInt(int64(numRegisteredBlsKeys)).Bytes(),
 				big.NewInt(nodeLimit).Bytes(),
 			},
 		}
 		v.eei.AddLogEntry(entry)
 	}
 
-	err = v.saveRegistrationData(args.CallerAddr, registrationData)
-	if err != nil {
-		v.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
-		return vmcommon.UserError
-	}
-
-	return vmcommon.Ok
 }
 
 func (v *validatorSC) activateStakingFor(
 	blsKeys [][]byte,
+	newKeys [][]byte,
 	registrationData *ValidatorDataV2,
 	fixedStakeValue *big.Int,
 	rewardAddress []byte,
 	ownerAddress []byte,
-) {
-	numRegistered := uint64(registrationData.NumRegistered)
+) bool {
+	numActivatedKey := uint64(registrationData.NumRegistered)
+
+	numRegisteredKeys := len(registrationData.BlsPubKeys)
+	if v.isNumberOfNodesTooHigh(numRegisteredKeys) {
+		return true
+	}
+
+	maxNumNodesToActivate := len(blsKeys)
+	if v.enableEpochsHandler.IsFlagEnabled(common.StakeLimitsFlag) {
+		maxNumNodesToActivate = v.computeNodeLimit() - numRegisteredKeys + len(newKeys)
+	}
+	nodesActivated := 0
+	if nodesActivated >= maxNumNodesToActivate && len(blsKeys) >= maxNumNodesToActivate {
```

Comment:
Is it ok that you
1. `nodesActivated := 0`
2. check `if nodesActivated >= maxNumNodesToActivate`
?

Reply:
yes. this is the case when you cannot activate any new node.